src/process/io/StreamSource.cpp

```diff
@@ -0,0 +1,268 @@
+        copiedBytes = bufLen;
+    }
+
+    if (numberOfBytes > rewoundBytes) {
+        copiedBytes = rewoundBytes;
+    }
+
+    // Copy copiedBytes of data from the right side!
+    if (targetBuffer != nullptr)
+        memcpy(targetBuffer, buf + bufLen - rewoundBytes, copiedBytes); // TODO
+    rewoundBytes -= copiedBytes;
+    currentPosition += copiedBytes;
+
+    delete[] buf;
+
+    return copiedBytes;
+}
+
+int64_t StreamSource::readFromStream(Bytef *targetBuffer, int numberOfBytes) {
+    char *readBuffer = new char[numberOfBytes]{0};
+    inputStream->sync();
+
+    inputStream->read(readBuffer, numberOfBytes);
+    streamsize amountRead = inputStream->gcount();
+
+    if (amountRead > 0) {
+        if (targetBuffer != nullptr)
+            memcpy(targetBuffer, readBuffer, amountRead);// TODO
+        rewindBuffer.emplace_back(readBuffer);
+        rewindBufferEntryLength.emplace_back(amountRead);
+        if (rewindBuffer.size() > maxSegmentsInBuffer) {
+            char *front = rewindBuffer.front();
+            rewindBuffer.pop_front();
+            rewindBufferEntryLength.pop_front();
+            delete[] front;
+        }
+        currentPosition += amountRead;          // Advance current position marker
+    } else {
+        delete[] readBuffer;
+    }
+    return amountRead;
+}
+
+int StreamSource::readChar() {
+    Byte result = 0;
+    int res = static_cast<int>(this->read(&result, 1));
+    return res < 0 ? res : (int) result;
+}
+
+/**
+ * A lot of cases apply here:
+ * - if absolute:
+ *   - nByte > curPos
+ *     => Skip to curPos or eof
+ *   - nByte < curPos
+ *     - nByte < sizeOfBuffer   => Go on
+ *     - nByte > sizeOfBuffer   => Error!
+ *
+ * - else if relative
+ *   - nByte > 0
+ *     => Skip by nByte
+ *   - nByte < 0
+ *     - nByte < sizeOfBuffer
+ *     - nByte > sizeOfBuffer
+ *
+ * To make things easier, convert a relative offset to an absolute offset first. So only the cases for absolute apply
+ *
+ * @param nByte The absolute position in the stream or the relative position to the current position
+ * @param absolute Absolute or relative position
+ * @return -1 for a failure or the new position in the stream
+ */
+int64_t StreamSource::seek(int64_t nByte, bool absolute) {
+    if (!absolute) {
+        nByte = currentPosition + nByte;
+    }
+
+    if (nByte > currentPosition) {
+        return skip(nByte);
+    } else {
+        int64_t difference = currentPosition - nByte;
+        return rewind(difference);
+    }
+}
+
+/**
+ * @param nByte
+ * @return
+ */
+int64_t StreamSource::rewind(int64_t nByte) {
+
+    // Make sure, rewind won't go over the border.
+    if (nByte > currentPosition) {
+        nByte = currentPosition;
+    }
+
+    // Check, if the rewind is too far
+    // We also need to be aware of a previous rewind!
+    if (nByte + rewoundBytes > getRewindBufferSize()) {
+        return -1; // We cannot go back further than the rewind buffer. This counts as an error
```

Comment:
Error return values should be at least documented in the method docu (better would again be something like "Optional".

Reply:
Again, see #44. You are right, this can be improved.